test/mongoose_rabbit_worker_SUITE.erl

```diff
@@ -0,0 +1,205 @@
+    ?assertNotMatch(ConnectionAndChannel0, ConnectionAndChannel1),
+
+    %% when amqp_call fails
+    gen_server:call(Worker, {amqp_call, {Exception, [ok]}}),
+
+    %% then
+    ConnectionAndChannel2 = get_worker_conn_and_chann(Worker),
+    ?assertNotMatch(ConnectionAndChannel1, ConnectionAndChannel2).
+
+
+worker_processes_msgs_when_queue_msg_len_limit_is_not_reached(Config) ->
+    %% given
+    Worker = proplists:get_value(worker_pid, Config),
+    Ref = make_ref(),
+    Lock = lock_fun(),
+    SendBack = send_back_fun(),
+
+    %% when
+    gen_server:cast(Worker, {amqp_publish, {Lock, [Ref]}, ok}),
+    gen_server:cast(Worker, {amqp_publish, {SendBack, [self(), Ref]}, ok}),
+    [gen_server:cast(Worker, {amqp_publish, ok, ok})
+     || _ <- lists:seq(1, ?MAX_QUEUE_LEN-1)],
+
+    %% unlock the worker
+    Worker ! Ref,
+
+    %% then
+    ?assertReceivedMatch(Ref, 100).
+
+worker_drops_msgs_when_queue_msg_len_limit_is_reached(Config) ->
+    %% given
+    Worker = proplists:get_value(worker_pid, Config),
+    Ref = make_ref(),
+    Lock = lock_fun(),
+    SendBack = send_back_fun(),
+
+    %% when
+    gen_server:cast(Worker, {amqp_publish, {Lock, [Ref]}, ok}),
+    gen_server:cast(Worker, {amqp_publish, {SendBack, [self(), Ref]}, ok}),
+    [gen_server:cast(Worker, {amqp_publish, ok, ok})
+     || _ <- lists:seq(1, ?MAX_QUEUE_LEN+1)],
+
+    %% unlock the worker
+    Worker ! Ref,
+
+    %% then
+    ?assertError({assertReceivedMatch_failed, _},
+                 ?assertReceivedMatch(Ref, 100)).
+
+%%--------------------------------------------------------------------
+%% Helpers
+%%--------------------------------------------------------------------
+
+mock_amqp() ->
+    [meck:new(M, [no_link, passthrough]) || M <- ?AMQP_MOCK_MODULES],
+    meck:expect(amqp_connection, start, fun(_) -> {ok, random_pid()} end),
+    meck:expect(amqp_connection, open_channel, fun(_) -> {ok, random_pid()} end),
+    meck:expect(amqp_connection, close, fun(_) -> ok end),
+    meck:expect(amqp_channel, close, fun(_) -> ok end),
+    meck:expect(amqp_channel, call,
+                fun(_, {F, A}) when is_function(F) -> apply(F, A);
+                   (_, _) -> ok
+                end),
+    meck:expect(amqp_channel, call,
+                fun(_, {F, A}, _) when is_function(F) -> apply(F, A);
+                   (_, _, _) -> ok
+                end).
+
+mock_metrics() ->
+    meck:new(mongoose_metrics, [no_link, passthrough]),
+    meck:expect(mongoose_metrics, update, fun(_, _, _) -> ok end).
+
+unload_amqp() ->
+    [meck:unload(M) || M <- ?AMQP_MOCK_MODULES].
+
+random_pid() ->
+    spawn(fun() -> ok end).
+
+lock_fun() ->
+    fun lock/1.
+
+lock(R) ->
+    receive
+        R -> unlocked
+    end.
+
+send_back_fun() ->
+    fun erlang:send/2.
+
+exception_fun() ->
+    fun exception/1.
+
+exception(_) ->
+    throw(exception).
+
+get_worker_conn_and_chann(Worker) ->
+    State = sys:get_state(Worker),
+    {element(3, State), element(4, State)}.
```

Comment:
oh, this code can break, if someone would edit the worker code.

One way to improve it is to implement the callback: 
http://erlang.org/doc/man/gen_server.html#Module:format_status-2

Which returns a map or a proplist of the record fields from it.
Of course, this callback on it's own would require its own tests :)

Reply:
Agree, that's naive implementation.

I can handle relevant test cases using `is_in_worker_state/2` implemented as follows:

```
is_in_worker_state(Worker, Elem) ->
    State = sys:get_state(Worker),
    lists:member(Elem, tuple_to_list(State)).
```

Does it make much sense now?

Reply:
Looks way better, IMO.

Reply:
Looks way better, IMO.